engine/apps/mattermost/tests/test_tasks.py

```diff
@@ -189,3 +214,290 @@ def test_on_alert_group_action_triggered_async_failure(
+    user_notification_policy = make_user_notification_policy(
+        user=user,
+        step=UserNotificationPolicy.Step.NOTIFY,
+        notify_by=UserNotificationPolicy.NotificationChannel.TESTONLY,
+    )
+
+    url = "{}/api/v4/posts".format(settings.MATTERMOST_HOST)
+    data = make_mattermost_post_response()
+    mock_response = httpretty.Response(json.dumps(data), status=status.HTTP_200_OK)
+    httpretty.register_uri(httpretty.POST, url, responses=[mock_response])
+
+    notify_user_about_alert_async(
+        user_pk=user.pk, alert_group_pk=alert_group.pk, notification_policy_pk=user_notification_policy.pk
+    )
+
+    log_record = user_notification_policy.personal_log_records.last()
+    mattermost_message = alert_group.mattermost_messages.order_by("created_at").last()
+    assert mattermost_message.post_id == data["id"]
+    assert mattermost_message.channel_id == data["channel_id"]
+    assert mattermost_message.message_type == MattermostMessage.USER_NOTIFACTION_MESSAGE
+    assert (
+        log_record.notification_error_code
+        == UserNotificationPolicyLogRecord.ERROR_NOTIFICATION_IN_MATTERMOST_USER_NOT_IN_MATTERMOST
+    )
+
+
+@pytest.mark.django_db
+@httpretty.activate(verbose=True, allow_net_connect=False)
+@pytest.mark.parametrize("status_code", [400, 401])
+def test_notify_user_about_alert_async_api_failure(
+    make_organization_and_user,
+    make_alert_receive_channel,
+    make_alert_group,
+    make_alert,
+    make_mattermost_channel,
+    make_mattermost_message,
+    make_mattermost_post_response_failure,
+    make_user_notification_policy,
+    make_mattermost_user,
+    status_code,
+):
+    organization, user = make_organization_and_user()
+    alert_receive_channel = make_alert_receive_channel(organization)
+    alert_group = make_alert_group(alert_receive_channel)
+    make_alert(alert_group=alert_group, raw_request_data=alert_receive_channel.config.example_payload)
+    make_mattermost_channel(organization=organization, is_default_channel=True)
+    make_mattermost_message(alert_group, MattermostMessage.ALERT_GROUP_MESSAGE)
+    make_mattermost_user(user=user)
+    user_notification_policy = make_user_notification_policy(
+        user=user,
+        step=UserNotificationPolicy.Step.NOTIFY,
+        notify_by=UserNotificationPolicy.NotificationChannel.TESTONLY,
+    )
+
+    url = "{}/api/v4/posts".format(settings.MATTERMOST_HOST)
+    data = make_mattermost_post_response_failure(status_code=status_code)
+    mock_response = httpretty.Response(json.dumps(data), status=status_code)
+    httpretty.register_uri(httpretty.POST, url, status=status_code, responses=[mock_response])
+
+    if status_code != 401:
+        with pytest.raises(MattermostAPIException):
+            notify_user_about_alert_async(
+                user_pk=user.pk, alert_group_pk=alert_group.pk, notification_policy_pk=user_notification_policy.pk
+            )
+    else:
```

Comment:
What should happen in each case? (401, no retry, log; other, retry, no log?)

Reply:
Yup that is the idea